fcrepo4/fcrepo4
# FCREPO-3200 - Default Digests

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3200

# What does this Pull Request do?
* Internal fedora binaries will always record a default digest
* Adds a repository wide default digest algorithm from a system property.
   * Can only be SHA512 or SHA256 due to OCFL restrictions
* New OCFL objects are configured to use this default digest algorithm
* Added aliases for digest algorithms to help with matching across domains
* Adds integration tests for binary persistence in OCFL


# How should this be tested?

Digests are not currently visible in http responses unfortunately, so you would need to look at the OCFL persistence to verify results:

```
curl -XPOST http://localhost:8080/rest/ -H"Slug: binary_default_digest" --data-binary @README.md -H"Content-type: text/plain" -ufedoraAdmin:fedoraAdmin

# verify that the sha512 is set (adjust for your OCFL storage root)
less /var/folders/v6/r6tq3p3n2h165cw93s39rl25cv0v_w/T/fcrepo.ocfl.storage.root.dir/84/43/86/binary_default_digest/v1/content/.fcrepo/binary_default_digest.json 

{
  "parent": "info:fedora",
  "id": "info:fedora/binary_default_digest",
  "lastModifiedDate": "2020-06-03T18:14:13.987960Z",
  "interactionModel": "http://www.w3.org/ns/ldp#NonRDFSource",
  "createdBy": "fedoraAdmin",
  "createdDate": "2020-06-03T18:14:13.987960Z",
  "lastModifiedBy": "fedoraAdmin",
  "stateToken": "4BE570CA6FBA4D7ECD6D94DB8878B3C2",
  "contentSize": 5770,
  "digests": [
    "urn:sha-512:7003d24ed80964957eb59d590f8d63961a8db755bc5760b9d55c21d97d3b011497bce353e1ceaa9374efbe9f05f4b277010468b92a55c543b0dca787300f8681"
  ],
  "filename": "",
  "deleted": false,
  "mimeType": "text/plain",
  "archivalGroup": false,
  "objectRoot": true
}
```
```
# With bad digest
curl -XPOST http://localhost:8080/rest/ -H"Slug: binary_sha1_digest" --data-binary @LICENSE -H"Content-type: text/plain" -H"digest: sha=checksumdoesntmatch" -ufedoraAdmin:fedoraAdmin
# Checksum mismatch, computed SHA digest f5c0f25223b84f167e917701f91e90c49525def0 did not match expected value checksumdoesntmatch

# With sha1 digest
curl -XPOST http://localhost:8080/rest/ -H"Slug: binary_sha1_digest" --data-binary @LICENSE -H"Content-type: text/plain" -H"digest: sha=f5c0f25223b84f167e917701f91e90c49525def0" -ufedoraAdmin:fedoraAdmin

# Verify that the sha512 and provided sha1 are present
less /var/folders/v6/r6tq3p3n2h165cw93s39rl25cv0v_w/T/fcrepo.ocfl.storage.root.dir/82/70/bb/binary_sha1_digest/v1/content/.fcrepo/binary_sha1_digest.json 

{
  "parent": "info:fedora",
  "id": "info:fedora/binary_sha1_digest",
  "lastModifiedDate": "2020-06-03T18:24:03.656085Z",
  "interactionModel": "http://www.w3.org/ns/ldp#NonRDFSource",
  "createdBy": "fedoraAdmin",
  "createdDate": "2020-06-03T18:24:03.656085Z",
  "lastModifiedBy": "fedoraAdmin",
  "stateToken": "9C8C83F606DBDA2BCB78A431E45AD3EC",
  "contentSize": 10831,
  "digests": [
    "urn:sha1:f5c0f25223b84f167e917701f91e90c49525def0",
    "urn:sha-512:05de5e29e68a7e33c86d494cf88da0c3c8c7c10ce5d3d678a548c2538991f806728d6e44495d3b59da13d279533c29952f6467b61a57415e9239d76e23201b22"
  ],
  "filename": "",
  "deleted": false,
  "mimeType": "text/plain",
  "archivalGroup": false,
  "objectRoot": true
}
```

# Notes
The SHA512 is generated both when the file is written to staging, and the OCFL client also generates and verifies it. Some efficiency should be gained later in FCREPO-3201 by passing the digest along with the OCFL client in terms of transmission fixity checking.

# Interested parties
@awoods @pwinckles
